File: test-junit5/src/test/java/io/micronaut/test/junit5/ApplicationRunAnotherTest.java

package io.micronaut.test.junit5;

import io.micronaut.context.BeanContext;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.client.HttpClient;
import io.micronaut.http.client.annotation.Client;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.micronaut.test.annotation.MockBean;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import jakarta.inject.Inject;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@MicronautTest
class ApplicationRunAnotherTest {

    @Inject
    @Client("/")
    HttpClient client;

    @Inject
    TestService testService;

    @Inject
    BeanContext beanContext;

    @Test
    void testPingServer() {
        when(testService.doStuff()).thenReturn("mocked by " + ApplicationRunAnotherTest.class.getName());
        Assertions.assertEquals(
                "mocked by " +ApplicationRunAnotherTest.class.getName(),
                client.toBlocking().retrieve(HttpRequest.GET("/test"), String.class)
        );
        verify(testService).doStuff();
    }

    @Test
    void testPingServerAgain() {
        when(testService.doStuff()).thenReturn("changed by " + ApplicationRunAnotherTest.class.getName());
        Assertions.assertEquals(
                "changed by " +ApplicationRunAnotherTest.class.getName(),
                client.toBlocking().retrieve(HttpRequest.GET("/test"), String.class)
        );
        verify(testService).doStuff();
    }

    @MockBean(DefaultTestService.class)
    TestService testService() {
        return mock(TestService.class);
    }
}

File: test-kotest5/src/test/kotlin/io/micronaut/test/kotest5/ApplicationRunAnotherTest.kt

package io.micronaut.test.kotest5

import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import io.micronaut.http.HttpRequest
import io.micronaut.http.client.HttpClient
import io.micronaut.http.client.annotation.Client
import io.micronaut.test.extensions.kotest5.annotation.MicronautTest
import io.micronaut.test.annotation.MockBean
import io.micronaut.test.extensions.kotest5.MicronautKotest5Extension.getMock
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify

@MicronautTest
class ApplicationRunAnotherTest(
        @Client("/") private val client: HttpClient,
        private val testService: io.micronaut.test.kotest5.TestService
): BehaviorSpec({

    val specName = javaClass.simpleName

    given("testPingServer") {
        `when`("the mock is setup") {
            val mock = getMock(testService)

            every { mock.doStuff() } returns "mocked by $specName"

            val response = client.toBlocking().retrieve(
                    HttpRequest.GET<String>("/test"), String::class.java)

            then("the response comes from the mock") {
                response shouldBe "mocked by ApplicationRunAnotherTest"
                verify {
                    mock.doStuff()
                }
            }
        }

        `when`("the mock is changed") {
            val mock = getMock(testService)

            every { mock.doStuff() } returns "changed by $specName"

            val response = client.toBlocking().retrieve(
                    HttpRequest.GET<String>("/test"), String::class.java)

            then("the mock was changed") {
                response shouldBe "changed by ApplicationRunAnotherTest"
                verify {
                    mock.doStuff()
                }
            }
        }
    }

}) {

    @MockBean(io.micronaut.test.kotest5.DefaultTestService::class)
    fun testService(): io.micronaut.test.kotest5.TestService {
        return mockk()
    }
}
